flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java

```diff
@@ -98,6 +104,15 @@ public class TaskLocalStateStoreImpl implements OwnedTaskLocalStateStore {
     @GuardedBy("lock")
     private final SortedMap<Long, TaskStateSnapshot> storedTaskStateByCheckpointID;
 
+    /** The relationship between checkpoints. (cp1,cp3) means cp3 refer to cp1's state. */
+    private final Map<Long, Long> referredByCheckpointID;
+
+    /**
+     * (MaterializationID, created by checkpointID). Because local store only keeps one checkpoint,
+     * at most one materialization can exist at the same time.
+     */
+    private Tuple2<Long, Long> currentMaterializationID;
```

Comment:
`lastChangelogMaterializationIdAndCheckpoint`?

Or maybe these two new fields should form a new helper class encapsulating changelog-specific logic? 
And then probably tuple2 can be replaced with just two fields?

I think it should be more clear and easier to test. WDYT?

Reply:
Nice suggestion, I created `ChangelogTaskLocalStateStore` to encapsulate changelog-specific logic.